VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol

```diff
@@ -0,0 +1,1098 @@
+                params,
+                nodeInfo.regionCode,
+                nodeInfo.datacenterId);
+                
+            params = abi.encodePacked(
+                params,
+                nodeInfo.lat,
+                nodeInfo.long,
+                nodeInfo.slotCount
+            );
+
+            verifyRequest(nodeInfo.didAddress, params, requestSignature, requestProof);
+
+            verifyAuthSignature(nodeInfo.didAddress, authSignature);
+        }
+
+        if (_slotInfo.isStakingRequired) {
+            stakeToken(nodeInfo.didAddress, tx.origin, nodeInfo.slotCount);
+        }
+
+        storeNodeInfo(nodeInfo);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function removeNodeStart(
+        address didAddress,
+        uint unregisterDateTime,
+        bytes calldata requestSignature,
+        bytes calldata requestProof
+    ) external override {
+        uint nodeId = _didNodeId[didAddress];
+
+        // Check whether didAddress was registered before
+        if (nodeId == 0 || _nodeUnregisterTime[nodeId] != 0) {
+            revert InvalidDIDAddress();
+        }
+
+        if (unregisterDateTime < (block.timestamp + 28 days)) {
+            revert InvalidUnregisterTime();
+        }
+
+        bytes memory params = abi.encodePacked(didAddress, unregisterDateTime);
+        verifyRequest(didAddress, params, requestSignature, requestProof);
+
+        _nodeUnregisterTime[nodeId] = unregisterDateTime;
+
+        emit RemoveNodeStart(didAddress, unregisterDateTime);
+    }
+
+    /**
+     * @notice Release staked token to the requestor of `removeNodeComplete()` function
+     * @dev Internal function. Called inside of `removeNodeComplete()` function
+     * @param didAddress DID address that calls the `removeNodeComplete()` function
+     * @param to EOA wallet that will received the released token
+     */
+    function releaseToken(address didAddress, address to) internal {
+        uint totalAmount = _stakedTokenAmount[didAddress];
+
+        if (totalAmount != 0) {
+            IERC20Upgradeable(vdaTokenAddress).transfer(to, totalAmount);
+            _stakedTokenAmount[didAddress] = 0;
+        }        
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function removeNodeComplete(
+        address didAddress,
+        bytes calldata requestSignature,
+        bytes calldata requestProof
+    ) external override {
+        uint nodeId = _didNodeId[didAddress];
+        {
+            if (nodeId == 0) {
+                revert InvalidDIDAddress();
+            }
+
+            if (_nodeUnregisterTime[nodeId] == 0 || _nodeUnregisterTime[nodeId] > block.timestamp) {
+                revert InvalidUnregisterTime();
+            }
+
+            bytes memory params = abi.encodePacked(didAddress);
+            verifyRequest(didAddress, params, requestSignature, requestProof);
+        }
+
+        // Release staked token
+        releaseToken(didAddress, tx.origin);
```

Comment:
Just a thought. Perhaps we don't automatically unstake the tokens when `removeNodeComplete()` and instead they have to be removed via `withdraw()`? In that way it's possible to remove a node and then add another node, without unstaking / restaking.

What do you think?

Reply:
Good idea! Here is my thought.
By the way, we can't be sure that the staked amount is enough for the user when they add a node again.
If it's not enough, they should deposit token. This can be by various reasons. For example, the `slotCount` in the user's input might be different from the previous one. Or the contract owner might change the StakePerSlot.
In case, if a user has some staked tokens in the contract, he should calculate how many tokens he should deposit more to add a new node. If there is no staked tokens, the user can calculate the required tokens so simply by `STAKE_PER_SLOT * SLOT_COUNT`. But if he has some staked tokens in the contract, he should calculate required tokens a bit complicated. He should ask the contract how many tokens are remaining and then substract this remaining tokens from the total required tokens. This might be unconveninent for users.

Also, some users don't add node and want to release the staked tokens when they comple remove operation.

And this is another reason. If users should deposit token in `addNode`, it'd be ought to release the token in the `removeNodeComplete`.

Reply:
Thanks for your thoughts.

So you suggest leaving the functionality as it current is?

Reply:
Yes, let's leave it as now.